_remake/lib/init-service-routes.js

```diff
@@ -0,0 +1,198 @@
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const multer = require('multer');
+const extract = require('extract-zip');
+const shell = require('shelljs');
+
+// configure multer storage
+const storage = multer.diskStorage({
+  destination: (req, file, cb) => {
+    cb(null, global.config.location.tmp);
+  },
+  filename: (req, file, cb) => {
+    cb(null, file.originalname);
+  }
+})
+
+const upload = multer({ storage });
+
+// check JWT token received in the header of protected requests
+function checkJWT(req, res, next) {
+  const bearerHeader = req.headers["authorization"]; // get token
+  if (typeof bearerHeader !== 'undefined') {
+    const bearer = bearerHeader.split(" ")[1];
+    // verity received token against the JWT secret used for generating it
+    jwt.verify(bearer, config.jwt.secret, (err, data) => {
+      if (err) {
+        return res.status(403).json({ message: 'Bad request: ' + err }).end();
+      } else {
+        req.user_id = data.id;
+        next();
+      }
+    })
+  } else {
+    return res.status(403).json({ message: 'Bad request: authorization header not provided' }).end();
+  }
+}
+
+// validate email
+function validEmail(req, res, next) {
+  // regex source: https://stackoverflow.com/a/46181
+  const emailRegex = /^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/
+  const email = req.body.email || req.query.email;
+  if (!email)
+    return res.status(400).json({ message : 'Bad request: email is missing.' }).end();
+  if (!emailRegex.test(email)) {
+    return res.status(400).json({ message : 'Bad request: email not valid.' }).end();
+  }
+  else next();
+}
+
+// validate password
+function validPass(req, res, next) {
+  // at least one digit and one letter + at least 8 chars long
+  const passRegex = /^(?=.*\d)(?=.*[a-z])[a-zA-Z0-9]{8,}$/
+  const password = req.body.password;
+  if (!password)
+    return res.status(400).json({ message : 'Bad request: password is missing.' }).end();
+  if (!passRegex.test(req.body.password))
+    return res.status(400).json({ message : 'Bad request: The password should be at least 8 characters long and it should contain at least one digit and a letter.' }).end();
+  else next();
+}
+
+// validate subdomain
+const validSubdomain = (req, res, next) => {
+  const subdomainRegex = /^[a-z]+[a-z0-9\-]*$/
+  const subdomain = req.query.subdomain || req.body.subdomain || req.body.appName;
```

Comment:
Because it's being send in different parts of request (url/body) and with different names (appName/subdomain) and I wanted to have only one validation function for all these cases.